### ISSUE TITLE
[QUESTION] no mssql-jdbc_auth-8.4.1.x64 in java.library.path, but it's on linux.

### ISSUE THREAD

**Issue by xaxis17** [microsoft/mssql-jdbc#1453]
## Question
Something is different when I got this error, this instance is running on a tomcat container "tomcat:jdk8-adoptopenjdk-hotspot", I followed multiple ways to copy mssql-jdbc_auth-8.4.1.x64.dll to local java.library.path which is below, but still I'm getting same error.

Which is weird that I copied the same dll on windows with same codes, it worked. The java.library.path suppose to be PATH on windows and LD_LIBRARY_PATH=/usr/local/tomcat/native-jni-lib on linux, I also printed path to confirm and they should be correct.

[https://i.stack.imgur.com/a5qLz.jpg](url)

[https://i.stack.imgur.com/3717c.jpg](url)

[https://i.stack.imgur.com/yQyQf.jpg](url)

## Relevant Issues and Pull Requests
https://github.com/microsoft/mssql-jdbc/issues/1439

I copied same thread from Stackoverflow.
https://stackoverflow.com/questions/64481219/no-mssql-jdbc-auth-8-4-1-x64-in-java-library-path

`Caused by: java.lang.UnsatisfiedLinkError: no mssql-jdbc_auth-8.4.1.x64 in java.library.path
    at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1860) ~[?:1.8.0_262]
    at java.lang.Runtime.loadLibrary0(Runtime.java:871) ~[?:1.8.0_262]
    at java.lang.System.loadLibrary(System.java:1124) ~[?:1.8.0_262]
    at com.microsoft.sqlserver.jdbc.AuthenticationJNI.<clinit>(AuthenticationJNI.java:51) ~[mssql-jdbc-8.4.1.jre8.jar:?]
    at com.microsoft.sqlserver.jdbc.SQLServerConnection.logon(SQLServerConnection.java:3942) ~[mssql-jdbc-8.4.1.jre8.jar:?]
    at com.microsoft.sqlserver.jdbc.SQLServerConnection.access$000(SQLServerConnection.java:85) ~[mssql-jdbc-8.4.1.jre8.jar:?]
    at com.microsoft.sqlserver.jdbc.SQLServerConnection$LogonCommand.doExecute(SQLServerConnection.java:3932) ~[mssql-jdbc-8.4.1.jre8.jar:?]
    at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:7375) ~[mssql-jdbc-8.4.1.jre8.jar:?]
    at com.microsoft.sqlserver.jdbc.SQLServerConnection.executeCommand(SQLServerConnection.java:3206) ~[mssql-jdbc-8.4.1.jre8.jar:?]
    at com.microsoft.sqlserver.jdbc.SQLServerConnection.connectHelper(SQLServerConnection.java:2713) ~[mssql-jdbc-8.4.1.jre8.jar:?]
    at com.microsoft.sqlserver.jdbc.SQLServerConnection.login(SQLServerConnection.java:2362) ~[mssql-jdbc-8.4.1.jre8.jar:?]
    at com.microsoft.sqlserver.jdbc.SQLServerConnection.connectInternal(SQLServerConnection.java:2213) ~[mssql-jdbc-8.4.1.jre8.jar:?]
    at com.microsoft.sqlserver.jdbc.SQLServerConnection.connect(SQLServerConnection.java:1276) ~[mssql-jdbc-8.4.1.jre8.jar:?]
    at com.microsoft.sqlserver.jdbc.SQLServerDriver.connect(SQLServerDriver.java:861) ~[mssql-jdbc-8.4.1.jre8.jar:?]
    at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:138) ~[HikariCP-3.4.2.jar:?]
    at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:354) ~[HikariCP-3.4.2.jar:?]
    at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:202) ~[HikariCP-3.4.2.jar:?]
    at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:473) ~[HikariCP-3.4.2.jar:?]
    at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:554) ~[HikariCP-3.4.2.jar:?]
    at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:115) ~[HikariCP-3.4.2.jar:?]
    at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:112) ~[HikariCP-3.4.2.jar:?]`

**Comment by rene-ye** [comment]
The .dll is used for Windows Integrated authentication and can't be used on Linux. Please use Kerberos or NTLM if you have a need for integrated authentication.

**Comment by xaxis17** [comment]
> 
> 
> The .dll is used for Windows Integrated Authentication and can't be used on Linux. Please use Kerberos or NTLM if you have a need for integrated authentication.

This explains, I tried to configure the /etc/krb5.conf for integration, but my domain is the aaa.com but my upn is bbb.com. My krb5.conf is like below, my configured user name is xxx@bbb.com, and then it warn me **Generic error (description in e-text) (60) - Unable to locate KDC for realm bbb.com**, I'm not quite sure about this.

`# Configuration snippets may be placed in this directory as well 
[logging]                                                        
 default = FILE:/var/log/krb5libs.log                            
 kdc = FILE:/var/log/krb5kdc.log                                 
 admin_server = FILE:/var/log/kadmind.log                        
                                                                 
[libdefaults]                                                    
 dns_lookup_realm = false                                        
 dns_lookup_kdc = true                                           
 ticket_lifetime = 24h                                           
 renew_lifetime = 7d                                             
 forwardable = true                                              
 default_realm = aaa.com                                    
                                                                 
[realms]                                                         
 aaa.com = {                                                
  kdc = s1812m920.aaa.com                                   
  admin_server = a1812m920.aaa.com                          
 }                                                               
                                                                 
[domain_realm]                                                   
 .aaa.com = aaa.com`

**Comment by xaxis17** [comment]
I think I solved this problem, but I'm not quite sure what's the background of it. the krb5.conf is still like I posted above, and connection is still with integratedSecurity=true;authenticationScheme=JavaKerberos;

I just remove the domain name from UPN and leave _**username**_ as username, not the xxx@aaa.com, not the aaa\xxx but only xxx.

Looks upn or backslash format could be searched in kerberos directly.
